## 1. The World Bank's international debt data
<p><img src="https://assets.datacamp.com/production/project_754/img/image.jpg" alt></p>
<p> Connecting to the database :The first line of code connects us to the <code>international_debt</code> database where the table <code>international_debt</code> is residing. Let's first <code>SELECT</code> <em>all</em> of the columns from the <code>international_debt</code> table. Also, we'll limit the output to the first ten rows to keep the output clean.</p>

In [2]:
%%sql
postgresql:///international_debt

'Connected: @international_debt'

In [25]:
%%sql
SELECT 
    *
FROM international_debt
LIMIT 10;

 * postgresql:///international_debt
10 rows affected.


[('Afghanistan', 'AFG', 'Disbursements on external debt, long-term (DIS, current US$)', 'DT.DIS.DLXF.CD', Decimal('72894453.700000003')),
 ('Afghanistan', 'AFG', 'Interest payments on external debt, long-term (INT, current US$)', 'DT.INT.DLXF.CD', Decimal('53239440.100000001')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (AMT, current US$)', 'DT.AMT.BLAT.CD', Decimal('61739336.899999999')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (DIS, current US$)', 'DT.DIS.BLAT.CD', Decimal('49114729.399999999')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (INT, current US$)', 'DT.INT.BLAT.CD', Decimal('39903620.100000001')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (AMT, current US$)', 'DT.AMT.MLAT.CD', Decimal('39107845')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (DIS, current US$)', 'DT.DIS.MLAT.CD', Decimal('23779724.300000001')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (INT, current US$)', 'DT.INT.MLAT.CD', Decimal('13335820')),
 ('Afghanistan', 'AFG', 'PPG, official creditors (AMT, current US$)', 'DT.AMT.OFFT.CD', Decimal('100847181.900000006')),
 ('Afghanistan', 'AFG', 'PPG, official creditors (DIS, current US$)', 'DT.DIS.OFFT.CD', Decimal('72894453.700000003'))]

## 2. Finding the number of distinct countries
<p>From the first ten rows, we can see the amount of debt owed by <em>Afghanistan</em> in the different debt indicators. But we do not know the number of different countries we have on the table. There are repetitions in the country names because a country is most likely to have debt in more than one debt indicator. </p>
<p>Thus, we are going to extract the number of unique countries present in the table. </p>

In [4]:
%%sql
SELECT 
    COUNT ( DISTINCT country_name )
FROM international_debt;

 * postgresql:///international_debt
1 rows affected.


[(124,)]

In [6]:
%%sql
SELECT 
    DISTINCT country_name
FROM international_debt
ORDER BY country_name;

 * postgresql:///international_debt
124 rows affected.


[('Afghanistan',),
 ('Albania',),
 ('Algeria',),
 ('Angola',),
 ('Armenia',),
 ('Azerbaijan',),
 ('Bangladesh',),
 ('Belarus',),
 ('Belize',),
 ('Benin',),
 ('Bhutan',),
 ('Bolivia',),
 ('Bosnia and Herzegovina',),
 ('Botswana',),
 ('Brazil',),
 ('Bulgaria',),
 ('Burkina Faso',),
 ('Burundi',),
 ('Cabo Verde',),
 ('Cambodia',),
 ('Cameroon',),
 ('Central African Republic',),
 ('Chad',),
 ('China',),
 ('Colombia',),
 ('Comoros',),
 ('Congo, Dem. Rep.',),
 ('Congo, Rep.',),
 ('Costa Rica',),
 ("Cote d'Ivoire",),
 ('Djibouti',),
 ('Dominica',),
 ('Dominican Republic',),
 ('Ecuador',),
 ('Egypt, Arab Rep.',),
 ('El Salvador',),
 ('Eritrea',),
 ('Eswatini',),
 ('Ethiopia',),
 ('Fiji',),
 ('Gabon',),
 ('Gambia, The',),
 ('Georgia',),
 ('Ghana',),
 ('Grenada',),
 ('Guatemala',),
 ('Guinea',),
 ('Guinea-Bissau',),
 ('Guyana',),
 ('Haiti',),
 ('Honduras',),
 ('IDA only',),
 ('India',),
 ('Indonesia',),
 ('Iran, Islamic Rep.',),
 ('Jamaica',),
 ('Jordan',),
 ('Kazakhstan',),
 ('Kenya',),
 ('Kosovo',),
 ('Kyrgyz Republic',),
 ('Lao PDR',),
 ('Least developed countries: UN classification',),
 ('Lebanon',),
 ('Lesotho',),
 ('Liberia',),
 ('Macedonia, FYR',),
 ('Madagascar',),
 ('Malawi',),
 ('Maldives',),
 ('Mali',),
 ('Mauritania',),
 ('Mauritius',),
 ('Mexico',),
 ('Moldova',),
 ('Mongolia',),
 ('Montenegro',),
 ('Morocco',),
 ('Mozambique',),
 ('Myanmar',),
 ('Nepal',),
 ('Nicaragua',),
 ('Niger',),
 ('Nigeria',),
 ('Pakistan',),
 ('Papua New Guinea',),
 ('Paraguay',),
 ('Peru',),
 ('Philippines',),
 ('Romania',),
 ('Russian Federation',),
 ('Rwanda',),
 ('Samoa',),
 ('Sao Tome and Principe',),
 ('Senegal',),
 ('Serbia',),
 ('Sierra Leone',),
 ('Solomon Islands',),
 ('Somalia',),
 ('South Africa',),
 ('South Asia',),
 ('Sri Lanka',),
 ('St. Lucia',),
 ('St. Vincent and the Grenadines',),
 ('Sudan',),
 ('Syrian Arab Republic',),
 ('Tajikistan',),
 ('Tanzania',),
 ('Thailand',),
 ('Timor-Leste',),
 ('Togo',),
 ('Tonga',),
 ('Tunisia',),
 ('Turkey',),
 ('Turkmenistan',),
 ('Uganda',),
 ('Ukraine',),
 ('Uzbekistan',),
 ('Vanuatu',),
 ('Venezuela, RB',),
 ('Vietnam',),
 ('Yemen, Rep.',),
 ('Zambia',),
 ('Zimbabwe',)]

## 3. Finding out the distinct debt indicators
<p>We can see there are a total of 124 countries present on the table. As we saw in the first section, there is a column called <code>indicator_name</code> that briefly specifies the purpose of taking the debt. Just beside that column, there is another column called <code>indicator_code</code> which symbolizes the category of these debts. Knowing about these various debt indicators will help us to understand the areas in which a country can possibly be indebted to. </p>

In [8]:
%%sql
SELECT
    COUNT(DISTINCT indicator_code)
FROM international_debt;

 * postgresql:///international_debt
1 rows affected.


[(25,)]

In [7]:
%%sql
SELECT
    DISTINCT indicator_code
FROM international_debt;

 * postgresql:///international_debt
25 rows affected.


[('DT.AMT.PCBK.CD',),
 ('DT.DIS.DLXF.CD',),
 ('DT.INT.DLXF.CD',),
 ('DT.INT.DPNG.CD',),
 ('DT.AMT.DLXF.CD',),
 ('DT.DIS.PCBK.CD',),
 ('DT.AMT.PROP.CD',),
 ('DT.AMT.MLAT.CD',),
 ('DT.AMT.OFFT.CD',),
 ('DT.INT.OFFT.CD',),
 ('DT.INT.MLAT.CD',),
 ('DT.DIS.MLAT.CD',),
 ('DT.INT.PRVT.CD',),
 ('DT.INT.PBND.CD',),
 ('DT.AMT.BLAT.CD',),
 ('DT.DIS.OFFT.CD',),
 ('DT.INT.PCBK.CD',),
 ('DT.INT.PROP.CD',),
 ('DT.AMT.PRVT.CD',),
 ('DT.DIS.PROP.CD',),
 ('DT.DIS.PRVT.CD',),
 ('DT.INT.BLAT.CD',),
 ('DT.AMT.PBND.CD',),
 ('DT.DIS.BLAT.CD',),
 ('DT.AMT.DPNG.CD',)]

## 4. Totaling the amount of debt owed by the countries
<p>As mentioned earlier, the financial debt of a particular country represents its economic state. But if we were to project this on an overall global scale, how will we approach it?</p>
<p>Let's switch gears from the debt indicators now and find out the total amount of debt (in USD) that is owed by the different countries. This will give us a sense of how the overall economy of the entire world is holding up.</p>

In [5]:
%%sql
SELECT 
    ROUND(SUM(debt)/1000000, 2) as global_total_debt_in_millions
FROM international_debt; 

 * postgresql:///international_debt
1 rows affected.


[(Decimal('3079734.49'),)]

## 5. Country with the highest debt
<p>Now that we have the exact total of the amounts of debt owed by several countries, let's now find out the country that owns the highest amount of debt along with the amount. <strong>Note</strong> that this debt is the sum of different debts owed by a country across several categories. This will help to understand more about the country in terms of its socio-economic scenarios. </p>

In [23]:
%%sql
SELECT 
    country_name,
    ROUND(SUM(debt)/1000000, 2) as total_debt_in_millions
FROM international_debt
GROUP BY country_name
ORDER BY total_debt_in_millions DESC
LIMIT 1; 

 * postgresql:///international_debt
1 rows affected.


[('China', Decimal('285793.49'))]

## 6. Average amount of debt across indicators
<p>So, it was <em>China</em>. A more in-depth breakdown of China's debts can be found <a href="https://datatopics.worldbank.org/debt/ids/country/CHN">here</a>. </p>
<p>We now have a brief overview of the dataset and a few of its summary statistics. We already have an idea of the different debt indicators in which the countries owe their debts. We can dig even further to find out on an average how much debt a country owes? This will give us a better sense of the distribution of the amount of debt across different indicators.</p>

In [5]:
%%sql
SELECT 
    indicator_code AS debt_indicator,
    indicator_name,
    ROUND(AVG(debt)/1000000,2) AS average_debt_in_millions
FROM international_debt
GROUP BY debt_indicator, indicator_name
ORDER BY average_debt_in_millions DESC
LIMIT 10;

 * postgresql:///international_debt
10 rows affected.


[('DT.AMT.DLXF.CD', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('5904.87')),
 ('DT.AMT.DPNG.CD', 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', Decimal('5161.19')),
 ('DT.DIS.DLXF.CD', 'Disbursements on external debt, long-term (DIS, current US$)', Decimal('2152.04')),
 ('DT.DIS.OFFT.CD', 'PPG, official creditors (DIS, current US$)', Decimal('1958.98')),
 ('DT.AMT.PRVT.CD', 'PPG, private creditors (AMT, current US$)', Decimal('1803.69')),
 ('DT.INT.DLXF.CD', 'Interest payments on external debt, long-term (INT, current US$)', Decimal('1644.02')),
 ('DT.DIS.BLAT.CD', 'PPG, bilateral (DIS, current US$)', Decimal('1223.14')),
 ('DT.INT.DPNG.CD', 'Interest payments on external debt, private nonguaranteed (PNG) (INT, current US$)', Decimal('1220.41')),
 ('DT.AMT.OFFT.CD', 'PPG, official creditors (AMT, current US$)', Decimal('1191.19')),
 ('DT.AMT.PBND.CD', 'PPG, bonds (AMT, current US$)', Decimal('1082.62'))]

## 7. The highest amount of principal repayments
<p>We can see that the indicator <code>DT.AMT.DLXF.CD</code> tops the chart of average debt. This category includes repayment of long term debts. Countries take on long-term debt to acquire immediate capital. More information about this category can be found <a href="https://datacatalog.worldbank.org/principal-repayments-external-debt-long-term-amt-current-us-0">here</a>. </p>
<p>An interesting observation in the above finding is that there is a huge difference in the amounts of the indicators after the second one. This indicates that the first two indicators might be the most severe categories in which the countries owe their debts.</p>
<p>We can investigate this a bit more so as to find out which country owes the highest amount of debt in the category of long term debts (<code>DT.AMT.DLXF.CD</code>). Since not all the countries suffer from the same kind of economic disturbances, this finding will allow us to understand that particular country's economic condition a bit more specifically. </p>

In [7]:
%%sql
SELECT 
    country_name, 
    indicator_name
FROM international_debt
WHERE debt = (SELECT 
                 MAX(debt)
             FROM international_debt
             WHERE indicator_code = 'DT.AMT.DLXF.CD');

 * postgresql:///international_debt
1 rows affected.


[('China', 'Principal repayments on external debt, long-term (AMT, current US$)')]

## 8. The most common debt indicator
<p>China has the highest amount of debt in the long-term debt (<code>DT.AMT.DLXF.CD</code>) category. This is verified by <a href="https://data.worldbank.org/indicator/DT.AMT.DLXF.CD?end=2018&most_recent_value_desc=true">The World Bank</a>. It is often a good idea to verify our analyses like this since it validates that our investigations are correct. </p>
<p>We saw that long-term debt is the topmost category when it comes to the average amount of debt. But is it the most common indicator in which the countries owe their debt? Let's find that out. </p>

In [11]:
%%sql
SELECT 
    indicator_code,
    COUNT(indicator_code) AS indicator_count
FROM international_debt
GROUP BY indicator_code
ORDER BY indicator_count DESC, indicator_code DESC
LIMIT 20;

 * postgresql:///international_debt
20 rows affected.


[('DT.INT.OFFT.CD', 124),
 ('DT.INT.MLAT.CD', 124),
 ('DT.INT.DLXF.CD', 124),
 ('DT.AMT.OFFT.CD', 124),
 ('DT.AMT.MLAT.CD', 124),
 ('DT.AMT.DLXF.CD', 124),
 ('DT.DIS.DLXF.CD', 123),
 ('DT.INT.BLAT.CD', 122),
 ('DT.DIS.OFFT.CD', 122),
 ('DT.AMT.BLAT.CD', 122),
 ('DT.DIS.MLAT.CD', 120),
 ('DT.DIS.BLAT.CD', 113),
 ('DT.INT.PRVT.CD', 98),
 ('DT.AMT.PRVT.CD', 98),
 ('DT.INT.PCBK.CD', 84),
 ('DT.AMT.PCBK.CD', 84),
 ('DT.INT.DPNG.CD', 79),
 ('DT.AMT.DPNG.CD', 79),
 ('DT.INT.PBND.CD', 69),
 ('DT.AMT.PBND.CD', 69)]

In [21]:
%%sql
SELECT 
    indicator_code,
    COUNT(indicator_code) AS indicator_count
FROM international_debt
GROUP BY indicator_code
HAVING COUNT(indicator_code) = (
    SELECT COUNT(DISTINCT country_name)
    FROM international_debt
)
ORDER BY indicator_code ASC;

 * postgresql:///international_debt
6 rows affected.


[('DT.AMT.DLXF.CD', 124),
 ('DT.AMT.MLAT.CD', 124),
 ('DT.AMT.OFFT.CD', 124),
 ('DT.INT.DLXF.CD', 124),
 ('DT.INT.MLAT.CD', 124),
 ('DT.INT.OFFT.CD', 124)]

## 9. Other viable debt issues and conclusion
<p>There are a total of six debt indicators in which all the countries listed in our dataset have taken debt. The indicator <code>DT.AMT.DLXF.CD</code> is also there in the list. So, this gives us a clue that all these countries are suffering from a common economic issue. But that is not the end of the story, but just a part of the story.</p>
<p>Let's change tracks from <code>debt_indicator</code>s now and focus on the amount of debt again. Let's find out the maximum amount of debt that each country has. With this, we will be in a position to identify the other plausible economic issues a country might be going through.</p>

In [24]:
%%sql
SELECT 
    country_name, 
    ROUND(MAX(debt)/1000000,2) AS maximum_debt_in_millions
FROM international_debt
GROUP BY country_name
ORDER BY maximum_debt_in_millions DESC
LIMIT 10;

 * postgresql:///international_debt
10 rows affected.


[('China', Decimal('96218.62')),
 ('Brazil', Decimal('90041.84')),
 ('Russian Federation', Decimal('66589.76')),
 ('Turkey', Decimal('51555.03')),
 ('South Asia', Decimal('48756.30')),
 ('Least developed countries: UN classification', Decimal('40160.77')),
 ('IDA only', Decimal('34531.19')),
 ('India', Decimal('31923.51')),
 ('Indonesia', Decimal('30916.11')),
 ('Kazakhstan', Decimal('27482.09'))]